### DISCUSSION TITLE
Что будет и чего не будет в эмуляторе

### DISCUSSION THREAD

**Discussion by ogamespec** [emu-russia/breaknes#181] [Announcements]
Эмулятор ещё не готов, поэтому можно сразу написать чего не будет.

- Save States. Не нужно, играйте без них. Используйте виртуальную машину и песочницу, в которой запущен эмулятор.
- AVI Record. Нах. Используйте программы захвата
- TAS. Есть уже дофига эмуляторов для TAS
- Удобный отладчик. Используйте Mesen. В эмуляторе будут базовые отладочные механизмы, чтобы проверить что там что-то шевелится. Всё остальное мы проверяем специализированными тулзами (Breaks Debugger, PPU Player) и юнит-тестами
- Frame Skip. Это что?
- Перемотка. Используйте виртуальную машину и песочницу, в которой запущен эмулятор.
- Скорости. Пока на первое время 1-2 FPS будет достаточно
- Запись аудио. Используйте программы захвата
- Game Genie. Нах нужно.

Короче видно что эмулятор будет максимально казуальным и примерно соответствует реальной консоли. Что умеет консоль, то умеет и эмулятор. Примерно так.

Киллер фичи.

- Точная симуляция основных чипов (собственно основная суть всей затеи)
- Настраиваемые борды. Можно слепить свою борду.
- @eugene-s-nesdev попросил также сделать возможность лепить химеру из чипов, т.к. нет ни одного полностью нормального денди-чипа, поэтому лучше всего будет взять по кускам из разных и слепить свой

Потом дополню.

**Comment by eugene-s-nesdev** [comment]
Дополню последний пункт, как я это вижу с позиции "перфекционизма". 
Эта модель близка к существующим "высокоуровневым" эмуляторам с денди-режимом.

Разработчики современных cycle-accurate эмуляторов (Mesen/puNES/BizHawk/Nintendulator/Nestopia)
берут за основу наработки по оригинальным Ricoh чипам, и прикручивают туда денди-тайминг.
Грубо говоря, в коде CPU/APU это выглядит как иной (/15) делитель частоты ЦП, а также "NTSC-подобные таблицы" APU
В коде PPU, это соответственно - иной Vblank, с более длинным пострендером.
https://forum.emu-russia.net/viewtopic.php?f=13&t=419
На выходе мы получаем довольно качественный и аутентичный денди-режим.

Проблема денди-чипов действительно состоит в том, что никакие из возможных комплектов не являются полностью беспроблемными. Все многокристальные UMC CPU имеют перепутанную скважность 25% и 50% на сквейр-каналах.
TA CPU лишены этого бага, зато там DPCM звучит заметно хуже, и сложно найти норм экземпляр. 
У всех PPUшек перепутаны местами красные и зеленые эмфаза-биты, и это наследие официального PAL NES, где такой же глюк.
Это лишь малая часть известных проблем. На некоторых тестах бларгга разные чипы выдают разные значения.
С однокристалками ситуация еще хуже. Там Triangle-канал заметно щёлкает, практически на всех моделях. 
Досконально все ревизии UM6561 не изучены, но точно известно, что А-ревизия довольно кривая (см. Prince of Persia)

Все вышеизложенное наталкивает на мысль воплотить в эмуляторе идеальную денди, по сути от нее нужны только тайминги. 
Глюки и особенности чипов могут быть опциональными.
